### PR TITLE
[camera][ios] add video codec option to expo camera

### LINKED ISSUE #8892: [expo-camera] Choose recording codec #Camera #HEVC
### iOS default: H.265/HEVC
Newer versions of iOS uses H.265/HEVC to generate video recording files by default.
This codec is not compatible with the majority of browsers, and they have no plans to support it in the future (https://caniuse.com/#feat=hevc).
This is a huge problem when an application should upload any kind of recording made in iOS that should be displayed in a web browser, or in any device that doesn't use this codec.

### Request
Since iOS 10 it is possible to specify the output format: https://developer.apple.com/documentation/avfoundation/avcapturemoviefileoutput/1388448-setoutputsettings?language=objc
Would be great to be able to specify the codec/output format of a video recorded by expo-camera.

### Informations (added on: June 26 2020)
RNCamera added this feature:
- https://github.com/react-native-community/react-native-camera/issues/1330
- https://github.com/bunchtogether/react-native-camera/commit/b970ea12e838e7eb55de5c5cc93c66b3ff69fc77

It could be helpful to read that.

### Optional improvements (added on: June 26 2020)
They also faced issues when setting an high quality with an older codec: https://github.com/react-native-community/react-native-camera/issues/1858
It seems they choose to throw an exception for that (I couldn't test it, correct me if I'm wrong).

### PR DESCRIPTION
**Pull request by ajsmth**
# Why

Closes #8892 

This PR adds a video codec option to the `recordAsync()` method. As mentioned in the issue above, the iOS camera defaults to recording videos in `H.265/HEVC`, however this is incompatible with almost all browsers, and so uploading these recordings for usage on the web is difficult without having an option to configure the encoding.

While this is specified as iOS only - I'm wondering if this should be enabled in Android as well - it looks like there is some kind of MediaRecorder API with options like this: https://developer.android.com/reference/android/media/MediaRecorder but I haven't done much research as of yet. Would love to know if anyone has tried this kind of thing in the past as it would be nice to keep parity between the two platforms

# How

Primarily used the `react-native-camera` implementation as a guide: https://github.com/react-native-camera/react-native-camera/issues/1330

The video codec enum is passed to the `recordAsync()` function, where it is mapped to the correct `AVVideoCodecType` enum that is eventually passed to the camera. When the recording ends, we cleanup this value so that it is not carried over into any other future invocations, and return the `uri` as well  as the `codec` used in the recording.

`recordAsync()` function now throws an error when an invalid codec is specified - this lets developers provide messages to their users detailing why the attempted recording didn't work.

It also adds a `Camera.getAvailableCodecsAsync()` function to let people provide some kind of validation when a specified codec is invalid for a given phone. 

# Test Plan

Since this uses the camera, it must be built on an actual device!

- passing an undefined codec to `recordAsync()` defaults to the normal codec as it was before (no change)
- providing a valid codec to `recordAsync()` results it it outputting a video encoded by whatever codec was passed in - this is returned from the `recordAsync()` function
- providing an invalid codec causes `recordAsync()` to reject with an error specifying that the codec passed in was invalid
- `Camera.getAvailableCodecsAsync()` returns a list of available codecs for a given device

If you are using the `sandbox` app you can use this gist to get an example up quickly:
https://gist.github.com/ajsmth/299d3530d55f95c65c59ecc4277b8c3e

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-camera/ios/EXCamera/EXCamera.m [hunk 1]
```diff
@@ -540,7 +540,33 @@ - (void)record:(NSDictionary *)options resolve:(UMPromiseResolveBlock)resolve re
...
+            [self cleanupMovieFileCapture];
+            _videoRecordedResolve = nil;
+            _videoRecordedReject = nil;
+            
+            return;
+        }
+    }
+     
+    
+
```
Comment (bbarthec) [inline packages/expo-camera/ios/EXCamera/EXCamera.m:569]:
```suggestion
    if (options[@"codec"]) {
      AVVideoCodecType videoCodecType = [EXCameraUtils videoCodecForType: [options[@"codec"] integerValue]];

      if ([self.movieFileOutput.availableVideoCodecTypes containsObject:videoCodecType]) {
        [self.movieFileOutput setOutputSettings: @{AVVideoCodecKey: videoCodecType} forConnection: connection];    
        self.videoCodecType = videoCodecType;
      } else {
        NSString *videoCodecErrorMessage = [NSString stringWithFormat: @"Video Codec '%@' is not supported on this device", videoCodecType];
        reject(@"E_RECORDING_FAILED", videoCodecErrorMessage, nil);

        [self cleanupMovieFileCapture];
        _videoRecordedResolve = nil;
        _videoRecordedReject = nil;

        return;
      }
    }

```
We don't want some extra `warns` that pollutes the user-space. Rejecting with proper, descriptive error message is sufficient 😉

File: packages/expo-camera/ios/EXCamera/EXCamera.m [hunk 2]
```diff
@@ -871,12 +897,19 @@ - (void)captureOutput:(AVCaptureFileOutput *)captureOutput didFinishRecordingToO
...
   }
   if (success && _videoRecordedResolve != nil) {
-    _videoRecordedResolve(@{ @"uri": outputFileURL.absoluteString });
+      AVVideoCodecType videoCodec = self.videoCodecType;
+           
+      if (videoCodec == nil) {
+          videoCodec = [self.movieFileOutput.availableVideoCodecTypes firstObject];
+      }
+      
+      _videoRecordedResolve(@{ @"uri": outputFileURL.absoluteString, @"codec": videoCodec });
```
Comment (bbarthec) [inline packages/expo-camera/ios/EXCamera/EXCamera.m:906]:
```suggestion
    AVVideoCodecType videoCodec = self.videoCodecType;       
    if (!videoCodec) {
      videoCodec = [self.movieFileOutput.availableVideoCodecTypes firstObject];
    }  
    _videoRecordedResolve(@{ @"uri": outputFileURL.absoluteString, @"codec": videoCodec });
```

File: packages/expo-camera/ios/EXCamera/Utilities/EXCameraUtils.h [hunk 3]
```diff
@@ -21,7 +21,9 @@
 + (AVCaptureVideoOrientation)videoOrientationForDeviceOrientation:(UIDeviceOrientation)orientation;
 + (AVCaptureVideoOrientation)videoOrientationForInterfaceOrientation:(UIInterfaceOrientation)orientation;
 + (int)exportImageOrientation:(UIImageOrientation)orientation;
++(AVVideoCodecType) videoCodecForType: (EXCameraVideoCodec)videoCodecType;
```
Comment (bbarthec) [inline packages/expo-camera/ios/EXCamera/Utilities/EXCameraUtils.h:24]:
```suggestion
+ (AVVideoCodecType)videoCodecForType:(EXCameraVideoCodec)videoCodecType;
```

File: packages/expo-camera/ios/EXCamera/Utilities/EXCameraUtils.m [hunk 4]
```diff
@@ -97,6 +97,26 @@ + (int)exportImageOrientation:(UIImageOrientation)orientation
...
+        case EXCameraVideoCodecAppleProRes422:
+            return AVVideoCodecTypeAppleProRes422;
+        case EXCameraVideoCodecAppleProRes4444:
+            return AVVideoCodecTypeAppleProRes4444;
+        default:
+            return AVVideoCodecTypeHEVC;
+    }
+    
+}
+
```
Comment (bbarthec) [inline packages/expo-camera/ios/EXCamera/Utilities/EXCameraUtils.m:118]:
```suggestion
+ (AVVideoCodecType)videoCodecForType:(EXCameraVideoCodec)videoCodecType
{
  switch (videoCodecType) {
    case EXCameraVideoCodecH264:
      return AVVideoCodecTypeH264;
    case EXCameraVideoCodecHEVC:
      return AVVideoCodecTypeHEVC;
    case EXCameraVideoCodecJPEG:
      return AVVideoCodecTypeJPEG;
    case EXCameraVideoCodecAppleProRes422:
      return AVVideoCodecTypeAppleProRes422;
    case EXCameraVideoCodecAppleProRes4444:
      return AVVideoCodecTypeAppleProRes4444;
    default:
      return AVVideoCodecTypeHEVC;
  }
}

```

File: packages/expo-camera/ios/EXCamera/EXCameraManager.m [hunk 5]
```diff
@@ -350,4 +387,6 @@ + (NSDictionary *)pictureSizes
                                                                 reject:reject];
 }
 
+
+
```
Comment (bbarthec) [inline packages/expo-camera/ios/EXCamera/EXCameraManager.m:391]:
```suggestion
```

File: packages/expo-camera/ios/EXCamera/EXCameraManager.m [hunk 6]
```diff
@@ -330,6 +338,35 @@ + (NSDictionary *)pictureSizes
...
+    AVCaptureMovieFileOutput *movieFileOutput = [[AVCaptureMovieFileOutput alloc]  init];
+    
+    if ([session canAddOutput: movieFileOutput]) {
+        [session addOutput: movieFileOutput];
+    }
+    
+    resolve([movieFileOutput availableVideoCodecTypes]);
+}
+
+
```
Comment (bbarthec) [inline packages/expo-camera/ios/EXCamera/EXCameraManager.m:369]:
```suggestion
UM_EXPORT_METHOD_AS(getAvailableVideoCodecsAsync,
                    resolve:(UMPromiseResolveBlock)resolve
                    rejecter:(UMPromiseRejectBlock)reject)
{
  AVCaptureSession *session = [AVCaptureSession new];   
  [session beginConfiguration];

  NSError *error = nil;
  AVCaptureDevice *captureDevice = [EXCameraUtils deviceWithMediaType:AVMediaTypeVideo preferringPosition: AVCaptureDevicePositionFront];
  AVCaptureDeviceInput *captureDeviceInput = [AVCaptureDeviceInput deviceInputWithDevice:captureDevice error:&error];   
 
  if ([session canAddInput:captureDeviceInput]) {
     [session addInput:captureDeviceInput];
  }

  [session commitConfiguration];

  AVCaptureMovieFileOutput *movieFileOutput = [AVCaptureMovieFileOutput new];
  if ([session canAddOutput:movieFileOutput]) {
    [session addOutput:movieFileOutput];
  }
  
  resolve([movieFileOutput availableVideoCodecTypes]);
}

```

I'm curious about creating the whole AVCaptureSession here 🤔 
Aren't the lines
```
AVCaptureMovieFileOutput *movieFileOutput = [AVCaptureMovieFileOutput new];
resolve([movieFileOutput availableVideoCodecTypes]);
```
covering the whole functionality? 🤔 
I'm looking at the Apple Docs: https://developer.apple.com/documentation/avfoundation/avvideocodectype and https://developer.apple.com/documentation/avfoundation/avcapturemoviefileoutput/1648381-availablevideocodectypes
and wondering if these are not sufficient here? 
Or maybe I'm missing the point and we actually need to create the whole camera session to get the actual data 🤔

File: packages/expo-camera/src/Camera.tsx [hunk 7]
```diff
@@ -141,13 +151,12 @@ export default class Camera extends React.Component<CameraProps> {
     return await CameraManager.getAvailablePictureSizes(ratio, this._cameraHandle);
   }
 
-  async recordAsync(options?: CameraRecordingOptions): Promise<{ uri: string }> {
+  async recordAsync(options?: CameraRecordingOptions): Promise<{ uri: string; codec: VideoCodec }> {
```
Comment (bbarthec) [inline packages/expo-camera/src/Camera.tsx:154]:
```suggestion
  async recordAsync(options?: CameraRecordingOptions): Promise<{ uri: string }> {
```
I don't think we need to return this value here 🤔 
1. Use provide it explicitly, so it's known already
2. Default one is described in the docs? 🤔

**Review by bbarthec — CHANGES_REQUESTED** [review]
Pay attention to the code style/indentation. I know this iOS code style of ours is questionable a lot, but let's stick to it 🙏 
Left some other idea/suggestions in the code.

**Comment by ajsmth** [comment]
@bbarthec it looks like the CI deleted a bunch of your comments, sorry about that :\

The two I wanted to talk about were the `getAvailableVideoCodecsAsync()` - I was hoping the implementation would have been simple like docs suggest but after trial and error I had to create a session and add input/output to get that method to return the available codecs

The second was returning the codec from the `recordAsync()` - returning an optional codec string makes this method much easier to test, also the values from the enum are not exactly matching to what's returned from the device

File: packages/expo-camera/CHANGELOG.md [hunk 8]
```diff
@@ -5,6 +5,7 @@
 ### 🛠 Breaking changes
 
 ### 🎉 New features
+- On iOS added new `codec` parameter in `recordVideo()` method and new method `getAvailableVideoCodecsAsync()` that queries the device for available video codecs. ([#12772](https://github.com/expo/expo/pull/12772)) by [@ajsmth]
```
Comment (bbarthec) [inline packages/expo-camera/CHANGELOG.md:8]:
```suggestion
- On iOS added new `codec` parameter in `recordVideo()` method and new method `getAvailableVideoCodecsAsync()` that queries the device for available video codecs. ([#12772](https://github.com/expo/expo/pull/12772)) by [@ajsmth](https://github.com/ajsmth))

```

**Review by bbarthec — COMMENTED** [review]
Regarding these deleted comments - what a surprise we have here 😱 (I know the fix is coming though 😅)
Re `getAvailableVideoCodesAsync()` - Apple 🔔  shame 🔔  shame ... 🚶‍♂️ 
Re `recordAsync()` - TBH, I'm not convinced 🤔 Testing apart, but what does this sentence mean specifically?
> values from the enum are not exactly matching to what's returned from the device

Are we having different values on input (`method parameter`) and output (`optional codec string`)? 🤔


So, let me ask more specifically: does this optional codec return value bring any new information to the user? And by `any new information` I mean something that is not already known (either from docs or input values/params? 🤔)
If the answer is `yes` I can live with that optional codec string, other way I'm rather against the idea of introducing this, because why don't we provide similar return info regarding any other video parameter/meta-parameter? (`bitrate`, `length`, `dimensions`, etc.) 🤔

**Comment by ajsmth** [comment]
> Re `getAvailableVideoCodesAsync()` - Apple 🔔 shame 🔔 shame ... 🚶‍♂️

Hahah

> Re `recordAsync()` - TBH, I'm not convinced 🤔 Testing apart, but what does this sentence mean specifically?
> 
> > values from the enum are not exactly matching to what's returned from the device
> 
> Are we having different values on input (`method parameter`) and output (`optional codec string`)? 🤔
> 
> So, let me ask more specifically: does this optional codec return value bring any new information to the user? And by `any new information` I mean something that is not already known (either from docs or input values/params? 🤔)
> If the answer is `yes` I can live with that optional codec string, other way I'm rather against the idea of introducing this, because why don't we provide similar return info regarding any other video parameter/meta-parameter? (`bitrate`, `length`, `dimensions`, etc.) 🤔

What I was getting at it that we are passing in an enum and that is being converted to a string, however that string isn't any more valuable to the user than the enum would be, so you make a good point. I'm not sure the best way to test that the codec is being applied - any thoughts on this?

**Comment by bbarthec** [comment]
@ajsmth, 
1. > I'm not sure the best way to test that the codec is being applied - any thoughts on this?

Good question 🤔 I can only guess that we need to check meta-params of the recorded video, but I have no idea how we can achieve this 😞 
Ultimately, we can skip writing tests for this specific feature hoping it works as expected (or we can describe how to check it manually) 😅 

2. > do you think it's worth it to wrap some of these options in an async block regardless

If it doesn't break anything and possibly it improves/fixes something then refactoring for the win 🎉

**Comment by ajsmth** [comment]
@bbarthec 

> 1. > I'm not sure the best way to test that the codec is being applied - any thoughts on this?
> 
> Good question 🤔 I can only guess that we need to check meta-params of the recorded video, but I have no idea how we can achieve this 😞
> Ultimately, we can skip writing tests for this specific feature hoping it works as expected (or we can describe how to check it manually) 😅

I think the simplest way to do this is to return the codec used from the native side and restrict it in the JS implementation. Another option would be to provide a metadata argument similar to `exif` on the photo side wherein we could return the codec used and other data if needed

File: packages/expo-camera/ios/EXCamera/EXCamera.m [hunk 9]
```diff
@@ -540,7 +521,9 @@ - (void)record:(NSDictionary *)options resolve:(UMPromiseResolveBlock)resolve re
       [connection setPreferredVideoStabilizationMode:self.videoStabilizationMode];
     }
     [connection setVideoOrientation:[EXCameraUtils videoOrientationForDeviceOrientation:[[UIDevice currentDevice] orientation]]];
-
+    
+    [self setVideoOptions:options forConnection:connection onReject: reject];
```
Comment (brentvatne) [inline packages/expo-camera/ios/EXCamera/EXCamera.m:525]:
```suggestion
    [self setVideoOptions:options forConnection:connection onReject:reject];
```

File: packages/expo-camera/ios/EXCamera/EXCamera.m [hunk 10]
```diff
@@ -568,6 +551,54 @@ - (void)record:(NSDictionary *)options resolve:(UMPromiseResolveBlock)resolve re
   }
 }
 
+- (void)setVideoOptions:(NSDictionary *)options forConnection:(AVCaptureConnection *)connection onReject:(UMPromiseRejectBlock)reject
+{
+  UM_WEAKIFY(self);
+  dispatch_async(_sessionQueue, ^{
```
Comment (brentvatne) [inline packages/expo-camera/ios/EXCamera/EXCamera.m:557]: add comment indicating why this is invoked in async block

File: packages/expo-camera/ios/EXCamera/EXCameraManager.m [hunk 11]
```diff
@@ -330,6 +338,31 @@ + (NSDictionary *)pictureSizes
   resolve([[[self class] pictureSizes] allKeys]);
 }
 
+UM_EXPORT_METHOD_AS(getAvailableVideoCodecsAsync,
+                    resolve:(UMPromiseResolveBlock)resolve
```
Comment (brentvatne) [inline packages/expo-camera/ios/EXCamera/EXCameraManager.m:342]:
```suggestion
                    resolver:(UMPromiseResolveBlock)resolve
```

convention is to use `resolver` and `rejecter`

File: packages/expo-camera/src/Camera.types.ts [hunk 12]
```diff
@@ -94,6 +94,29 @@ export enum ImageType {
   jpg = 'jpg',
 }
 
+export enum VideoCodec {
```
Comment (brentvatne) [inline packages/expo-camera/src/Camera.types.ts:97]: let's add a docblock here about video codec to save @Simek the hassle of copying over from docs when he gets to this module for auto gen docs

File: scripts/git-hooks/post-checkout [hunk 13]
```diff
@@ -1,5 +1,3 @@
 #!/bin/sh
 command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/post-checkout.\n"; exit 2; }
 git lfs post-checkout "$@"
-
-./scripts/check-cocoapods-lockfiles.sh
```
Comment (brentvatne) [inline scripts/git-hooks/post-checkout:5]: revert

File: scripts/git-hooks/post-merge [hunk 14]
```diff
@@ -1,5 +1,3 @@
 #!/bin/sh
 command -v git-lfs >/dev/null 2>&1 || { echo >&2 "\nThis repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/post-merge.\n"; exit 2; }
 git lfs post-merge "$@"
-
-./scripts/check-cocoapods-lockfiles.sh
```
Comment (brentvatne) [inline scripts/git-hooks/post-merge:5]: revert

File: packages/jest-expo/src/preset/expoModules.js [hunk 15]
```diff
@@ -471,30 +470,41 @@ module.exports = {
...
-          { key: 0, argumentsCount: 0, name: 'getPermissionsAsync' },
-          { key: 1, argumentsCount: 0, name: 'logOutAsync' },
-          { key: 2, argumentsCount: 1, name: 'initializeAsync' },
-          { key: 3, argumentsCount: 1, name: 'setAdvertiserTrackingEnabledAsync' },
-          { key: 4, argumentsCount: 0, name: 'getAuthenticationCredentialAsync' },
-          { key: 5, argumentsCount: 1, name: 'setAutoLogAppEventsEnabledAsync' },
-          { key: 6, argumentsCount: 0, name: 'requestPermissionsAsync' },
-          { key: 7, argumentsCount: 1, name: 'logInWithReadPermissionsAsync' },
-          { key: 8, argumentsCount: 1, name: 'setAdvertiserIDCollectionEnabledAsync' },
+          { key: 0, argumentsCount: 0, name: 'flushAsync' },
```
Comment (brentvatne) [inline packages/jest-expo/src/preset/expoModules.js:489]: revert unrelated mocks changes, just update ExpoCameraManager for now

**Review by brentvatne — CHANGES_REQUESTED** [review]

**Review by bbarthec — APPROVED** [review]
Cleanup @brentvatne's suggestions and we're ready to merge 👍

**Review by bbarthec — APPROVED** [review]

**Review by brentvatne — APPROVED** [review]

**Comment by ajcpwnz** [comment]
Is there a way to use this locally?

**Pull request merged**